\section{tSNE analysis}

\subsection{Assesment of the tSNE perplexity hyperparameter}

\input{\tsneDir/tSNE.perplexity}
\clearpage

\subsection{tSNE parameters}

The parameters chosen for the subsequent tSNE plots are:

\begin{itemize}
\item Perplexity: \tSNEPerplexity
\item Maximum number of iterations: \tSNEMaxIter
\item Use fast approximation: \tSNEFast
\end{itemize}

\subsection{tSNE plots colored by factors of interest (e.g. by cluster)}

\input{\tsneDir/plot.rdims.tsne.factor}

\clearpage

\section{UMAP projection}

UMAP is a newer algorithm that often provides a superior projection to tSNE.

\input{\umapDir/plot.rdims.umap.factor}

\clearpage

\section{Numbers: plots of cell numbers by factors of interest (e.g. by cluster)}

\input{\groupNumbersDir/number.plots}

\clearpage

\section{Cluster relationships}

\subsection{Cluster distances in gene expression space}

The distances between the clusters was assessed using the ``BuildClusterTree'' function in the Seurat package, which ``constructs a phylogenetic tree relating the ``average'' cell from each identity class''.

\begin{figure}[H]
\includegraphics[width=1.0\textwidth,height=0.9\textheight,keepaspectratio]{{{\clusterDir/cluster.dendrogram}}}
\caption{Visualisation of inter-cluster distances (cluster average, gene-based)}
\end{figure}

\subsection{Cluster distances by cell pair-wise correlation in reduced dimension space}

An alternative distance metric is the median pairwise pearson correlation between the cells of each cluster in reduced dimension space (e.g. PCA).

\begin{figure}[H]
\includegraphics[width=1.0\textwidth,height=0.9\textheight,keepaspectratio]{{{\clusterDir/cluster.correlation.dendrogram}}}
\caption{Visualisation of inter-cluster distances (median pairwise pearson correlation in reduced dimension space)}
\end{figure}

The median pairwise pearson correlations between the cells of each cluster (reduced dimension space) are shown in the table below.

\input{\clusterDir/cluster.pairwise.correlations.tex}

\clearpage

\section{Cluster resolution analysis}

The \href{https://doi.org/10.1093/gigascience/giy083}{clustree algorithm} is used to compare the different clustering resolutions.

\begin{figure}[H]
\includegraphics[width=1.0\textwidth,height=0.9\textheight,keepaspectratio]{{{\clusterDir/clustree}}}
\caption{The relationships between the clusters identified at different resolutions}
\end{figure}

\clearpage